// yamapng/esm2015/lib/directives/marker.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, EventEmitter, Input, Output } from '@angular/core';
import { YaMapsAPIWrapper } from '../ya-maps-api-wrapper';
import { MarkerManager } from '../services/managers/marker-manager';
/** @type {?} */
let markerId = 0;
// tslint:disable-next-line:directive-class-suffix
export class YaMarker {
    /**
     * @param {?} _markerManager
     */
    constructor(_markerManager) {
        this._markerManager = _markerManager;
        this.draggable = false;
        this.preset = 'islands#blueIcon';
        this.markerClick = new EventEmitter();
        // tslint:disable-next-line:max-line-length
        this.dragEnd = new EventEmitter();
        this._markerAddedToManger = false;
        this._observableSubscriptions = [];
        this._id = (markerId++).toString();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (!this._markerAddedToManger) {
            this._markerManager.addMarker(this);
            this._markerAddedToManger = true;
            this._addEventListeners();
            return;
        }
        if (changes['showInfo']) {
            this._markerManager.showBalloon(this);
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._markerManager.deleteMarker(this);
        this._observableSubscriptions.forEach((s) => s.unsubscribe());
    }
    /**
     * @private
     * @return {?}
     */
    _addEventListeners() {
        // click event
        /** @type {?} */
        const cs = this._markerManager.createEventObservable('click', this).subscribe(() => {
            this._markerManager.showBalloon(this);
            this.markerClick.emit(null);
        });
        this._observableSubscriptions.push(cs);
        // dragend event
        // tslint:disable-next-line:max-line-length
        /** @type {?} */
        const ds = this._markerManager.createEventObservable('dragend', this).subscribe((e) => {
            /** @type {?} */
            const thisPlacemark = e.get('target');
            /** @type {?} */
            const coords = thisPlacemark.geometry.getCoordinates();
            this._markerManager.getNativeMarker(this).then((m) => {
                // tslint:disable-next-line:max-line-length
                this.dragEnd.emit((/** @type {?} */ ({ lat: coords[0], lng: coords[1], nativeMarker: m })));
            });
        });
        this._observableSubscriptions.push(ds);
    }
}
YaMarker.decorators = [
    { type: Directive, args: [{
                // tslint:disable-next-line:directive-selector
                selector: 'ya-marker',
                providers: [
                    YaMapsAPIWrapper
                ]
            },] }
];
/** @nocollapse */
YaMarker.ctorParameters = () => [
    { type: MarkerManager }
];
YaMarker.propDecorators = {
    latitude: [{ type: Input }],
    longitude: [{ type: Input }],
    balloonLayout: [{ type: Input }],
    balloonContentHeader: [{ type: Input }],
    balloonContentBody: [{ type: Input }],
    balloonContentFooter: [{ type: Input }],
    draggable: [{ type: Input }],
    preset: [{ type: Input }],
    iconContent: [{ type: Input }],
    showInfo: [{ type: Input }],
    iconLayout: [{ type: Input }],
    iconImageHref: [{ type: Input }],
    iconImageSize: [{ type: Input }],
    iconImageOffset: [{ type: Input }],
    markerClick: [{ type: Output }],
    dragEnd: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    YaMarker.prototype.latitude;
    /** @type {?} */
    YaMarker.prototype.longitude;
    /** @type {?} */
    YaMarker.prototype.balloonLayout;
    /** @type {?} */
    YaMarker.prototype.balloonContentHeader;
    /** @type {?} */
    YaMarker.prototype.balloonContentBody;
    /** @type {?} */
    YaMarker.prototype.balloonContentFooter;
    /** @type {?} */
    YaMarker.prototype.draggable;
    /** @type {?} */
    YaMarker.prototype.preset;
    /** @type {?} */
    YaMarker.prototype.iconContent;
    /** @type {?} */
    YaMarker.prototype.showInfo;
    /** @type {?} */
    YaMarker.prototype.iconLayout;
    /** @type {?} */
    YaMarker.prototype.iconImageHref;
    /** @type {?} */
    YaMarker.prototype.iconImageSize;
    /** @type {?} */
    YaMarker.prototype.iconImageOffset;
    /** @type {?} */
    YaMarker.prototype.markerClick;
    /** @type {?} */
    YaMarker.prototype.dragEnd;
    /**
     * @type {?}
     * @private
     */
    YaMarker.prototype._markerAddedToManger;
    /**
     * @type {?}
     * @private
     */
    YaMarker.prototype._id;
    /**
     * @type {?}
     * @private
     */
    YaMarker.prototype._observableSubscriptions;
    /**
     * @type {?}
     * @private
     */
    YaMarker.prototype._markerManager;
}
