// yamapng/esm5/lib/directives/claster.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input } from '@angular/core';
import { YaMapsAPIWrapper } from '../ya-maps-api-wrapper';
import { ClasterManager } from '../services/managers/claster-manager';
/** @type {?} */
var clasterId = 0;
var YaClaster = /** @class */ (function () {
    function YaClaster(_clasterManager) {
        this._clasterManager = _clasterManager;
        this._markerAddedToManger = false;
        this._id = (clasterId++).toString();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    YaClaster.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (!this._markerAddedToManger) {
            this._clasterManager.addClaster(this);
        }
    };
    /**
     * @return {?}
     */
    YaClaster.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () { };
    /**
     * @private
     * @return {?}
     */
    YaClaster.prototype._addEventListeners = /**
     * @private
     * @return {?}
     */
    function () { };
    YaClaster.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: 'ya-claster',
                    providers: [
                        YaMapsAPIWrapper
                    ]
                },] }
    ];
    /** @nocollapse */
    YaClaster.ctorParameters = function () { return [
        { type: ClasterManager }
    ]; };
    YaClaster.propDecorators = {
        markers: [{ type: Input }]
    };
    return YaClaster;
}());
export { YaClaster };
if (false) {
    /** @type {?} */
    YaClaster.prototype.markers;
    /**
     * @type {?}
     * @private
     */
    YaClaster.prototype._id;
    /**
     * @type {?}
     * @private
     */
    YaClaster.prototype._markerAddedToManger;
    /**
     * @type {?}
     * @private
     */
    YaClaster.prototype._clasterManager;
}
