// yamapng/esm5/lib/directives/marker.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, EventEmitter, Input, Output } from '@angular/core';
import { YaMapsAPIWrapper } from '../ya-maps-api-wrapper';
import { MarkerManager } from '../services/managers/marker-manager';
/** @type {?} */
var markerId = 0;
var YaMarker = /** @class */ (function () {
    function YaMarker(_markerManager) {
        this._markerManager = _markerManager;
        this.draggable = false;
        this.preset = 'islands#blueIcon';
        this.markerClick = new EventEmitter();
        // tslint:disable-next-line:max-line-length
        this.dragEnd = new EventEmitter();
        this._markerAddedToManger = false;
        this._observableSubscriptions = [];
        this._id = (markerId++).toString();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    YaMarker.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (!this._markerAddedToManger) {
            this._markerManager.addMarker(this);
            this._markerAddedToManger = true;
            this._addEventListeners();
            return;
        }
        if (changes['showInfo']) {
            this._markerManager.showBalloon(this);
        }
    };
    /**
     * @return {?}
     */
    YaMarker.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._markerManager.deleteMarker(this);
        this._observableSubscriptions.forEach(function (s) { return s.unsubscribe(); });
    };
    /**
     * @private
     * @return {?}
     */
    YaMarker.prototype._addEventListeners = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        // click event
        /** @type {?} */
        var cs = this._markerManager.createEventObservable('click', this).subscribe(function () {
            _this._markerManager.showBalloon(_this);
            _this.markerClick.emit(null);
        });
        this._observableSubscriptions.push(cs);
        // dragend event
        // tslint:disable-next-line:max-line-length
        /** @type {?} */
        var ds = this._markerManager.createEventObservable('dragend', this).subscribe(function (e) {
            /** @type {?} */
            var thisPlacemark = e.get('target');
            /** @type {?} */
            var coords = thisPlacemark.geometry.getCoordinates();
            _this._markerManager.getNativeMarker(_this).then(function (m) {
                // tslint:disable-next-line:max-line-length
                _this.dragEnd.emit((/** @type {?} */ ({ lat: coords[0], lng: coords[1], nativeMarker: m })));
            });
        });
        this._observableSubscriptions.push(ds);
    };
    YaMarker.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: 'ya-marker',
                    providers: [
                        YaMapsAPIWrapper
                    ]
                },] }
    ];
    /** @nocollapse */
    YaMarker.ctorParameters = function () { return [
        { type: MarkerManager }
    ]; };
    YaMarker.propDecorators = {
        latitude: [{ type: Input }],
        longitude: [{ type: Input }],
        balloonLayout: [{ type: Input }],
        balloonContentHeader: [{ type: Input }],
        balloonContentBody: [{ type: Input }],
        balloonContentFooter: [{ type: Input }],
        draggable: [{ type: Input }],
        preset: [{ type: Input }],
        iconContent: [{ type: Input }],
        showInfo: [{ type: Input }],
        iconLayout: [{ type: Input }],
        iconImageHref: [{ type: Input }],
        iconImageSize: [{ type: Input }],
        iconImageOffset: [{ type: Input }],
        markerClick: [{ type: Output }],
        dragEnd: [{ type: Output }]
    };
    return YaMarker;
}());
export { YaMarker };
if (false) {
    /** @type {?} */
    YaMarker.prototype.latitude;
    /** @type {?} */
    YaMarker.prototype.longitude;
    /** @type {?} */
    YaMarker.prototype.balloonLayout;
    /** @type {?} */
    YaMarker.prototype.balloonContentHeader;
    /** @type {?} */
    YaMarker.prototype.balloonContentBody;
    /** @type {?} */
    YaMarker.prototype.balloonContentFooter;
    /** @type {?} */
    YaMarker.prototype.draggable;
    /** @type {?} */
    YaMarker.prototype.preset;
    /** @type {?} */
    YaMarker.prototype.iconContent;
    /** @type {?} */
    YaMarker.prototype.showInfo;
    /** @type {?} */
    YaMarker.prototype.iconLayout;
    /** @type {?} */
    YaMarker.prototype.iconImageHref;
    /** @type {?} */
    YaMarker.prototype.iconImageSize;
    /** @type {?} */
    YaMarker.prototype.iconImageOffset;
    /** @type {?} */
    YaMarker.prototype.markerClick;
    /** @type {?} */
    YaMarker.prototype.dragEnd;
    /**
     * @type {?}
     * @private
     */
    YaMarker.prototype._markerAddedToManger;
    /**
     * @type {?}
     * @private
     */
    YaMarker.prototype._id;
    /**
     * @type {?}
     * @private
     */
    YaMarker.prototype._observableSubscriptions;
    /**
     * @type {?}
     * @private
     */
    YaMarker.prototype._markerManager;
}
