// yamapng/esm5/lib/directives/objectManager.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input } from '@angular/core';
import { YaMapsAPIWrapper } from '../ya-maps-api-wrapper';
import { ObjectManagerManager } from '../services/managers/objectManager-manager';
/** @type {?} */
var markerId = 0;
var YaObjectManager = /** @class */ (function () {
    function YaObjectManager(_manager) {
        this._manager = _manager;
        this.clusterize = false;
        this.clasterPreset = 'islands#blueIcon';
        this.objectPreset = 'islands#blueClusterIcons';
        this.gridSize = 0;
        this._observableSubscriptions = [];
        this._addedToManger = false;
        this._id = (markerId++).toString();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    YaObjectManager.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (!this._addedToManger) {
            this._manager.add(this);
            this._addedToManger = true;
            return;
        }
        if (changes['selectedObjectId']) {
            this._manager.navigateToGeoObject(this, this.selectedObjectId);
        }
        if (changes['filter']) {
            this._manager.setFilter(this, this.filter);
        }
    };
    YaObjectManager.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: 'ya-object-manager',
                    providers: [
                        YaMapsAPIWrapper
                    ]
                },] }
    ];
    /** @nocollapse */
    YaObjectManager.ctorParameters = function () { return [
        { type: ObjectManagerManager }
    ]; };
    YaObjectManager.propDecorators = {
        clusterize: [{ type: Input }],
        datasource: [{ type: Input }],
        clasterPreset: [{ type: Input }],
        objectPreset: [{ type: Input }],
        gridSize: [{ type: Input }],
        selectedObjectId: [{ type: Input }],
        filter: [{ type: Input }]
    };
    return YaObjectManager;
}());
export { YaObjectManager };
if (false) {
    /** @type {?} */
    YaObjectManager.prototype.clusterize;
    /** @type {?} */
    YaObjectManager.prototype.datasource;
    /** @type {?} */
    YaObjectManager.prototype.clasterPreset;
    /** @type {?} */
    YaObjectManager.prototype.objectPreset;
    /** @type {?} */
    YaObjectManager.prototype.gridSize;
    /** @type {?} */
    YaObjectManager.prototype.selectedObjectId;
    /** @type {?} */
    YaObjectManager.prototype.filter;
    /**
     * @type {?}
     * @private
     */
    YaObjectManager.prototype._id;
    /**
     * @type {?}
     * @private
     */
    YaObjectManager.prototype._observableSubscriptions;
    /**
     * @type {?}
     * @private
     */
    YaObjectManager.prototype._addedToManger;
    /**
     * @type {?}
     * @private
     */
    YaObjectManager.prototype._manager;
}
