// yamapng/esm2015/lib/services/managers/marker-manager.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, NgZone } from '@angular/core';
import { Observable } from 'rxjs';
import { YaMapsAPIWrapper } from '../../ya-maps-api-wrapper';
export class MarkerManager {
    /**
     * @param {?} _mapsWrapper
     * @param {?} _zone
     */
    constructor(_mapsWrapper, _zone) {
        this._mapsWrapper = _mapsWrapper;
        this._zone = _zone;
        this._markers = new Map();
    }
    /**
     * @param {?} marker
     * @return {?}
     */
    deleteMarker(marker) {
        /** @type {?} */
        const m = this._markers.get(marker);
        if (m == null) {
            return Promise.resolve();
        }
        // tslint:disable-next-line:no-shadowed-variable
        return m.then((m) => {
            return this._zone.run(() => {
                // tslint:disable-next-line:no-shadowed-variable
                this.getNativeMarker(marker).then((m) => {
                    this._mapsWrapper.removeGeo(m);
                    this._markers.delete(marker);
                });
            });
        });
    }
    /**
     * @param {?} marker
     * @return {?}
     */
    addMarker(marker) {
        /** @type {?} */
        const markerPromise = this._mapsWrapper.createMarker(marker);
        this._markers.set(marker, markerPromise);
    }
    /**
     * @param {?} marker
     * @return {?}
     */
    showBalloon(marker) {
        this.getNativeMarker(marker).then((m) => {
            m.balloon.open();
        });
    }
    /**
     * @param {?} marker
     * @return {?}
     */
    getNativeMarker(marker) {
        return this._markers.get(marker);
    }
    /**
     * @template T
     * @param {?} eventName
     * @param {?} marker
     * @return {?}
     */
    createEventObservable(eventName, marker) {
        return Observable.create((observer) => {
            this._markers.get(marker).then((m) => {
                m.events.add(eventName, (e) => this._zone.run(() => observer.next(e)));
            });
        });
    }
}
MarkerManager.decorators = [
    { type: Injectable }
];
/** @nocollapse */
MarkerManager.ctorParameters = () => [
    { type: YaMapsAPIWrapper },
    { type: NgZone }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    MarkerManager.prototype._markers;
    /**
     * @type {?}
     * @private
     */
    MarkerManager.prototype._mapsWrapper;
    /**
     * @type {?}
     * @private
     */
    MarkerManager.prototype._zone;
}
