// yamapng/esm5/lib/services/managers/marker-manager.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, NgZone } from '@angular/core';
import { Observable } from 'rxjs';
import { YaMapsAPIWrapper } from '../../ya-maps-api-wrapper';
var MarkerManager = /** @class */ (function () {
    function MarkerManager(_mapsWrapper, _zone) {
        this._mapsWrapper = _mapsWrapper;
        this._zone = _zone;
        this._markers = new Map();
    }
    /**
     * @param {?} marker
     * @return {?}
     */
    MarkerManager.prototype.deleteMarker = /**
     * @param {?} marker
     * @return {?}
     */
    function (marker) {
        var _this = this;
        /** @type {?} */
        var m = this._markers.get(marker);
        if (m == null) {
            return Promise.resolve();
        }
        // tslint:disable-next-line:no-shadowed-variable
        return m.then(function (m) {
            return _this._zone.run(function () {
                // tslint:disable-next-line:no-shadowed-variable
                _this.getNativeMarker(marker).then(function (m) {
                    _this._mapsWrapper.removeGeo(m);
                    _this._markers.delete(marker);
                });
            });
        });
    };
    /**
     * @param {?} marker
     * @return {?}
     */
    MarkerManager.prototype.addMarker = /**
     * @param {?} marker
     * @return {?}
     */
    function (marker) {
        /** @type {?} */
        var markerPromise = this._mapsWrapper.createMarker(marker);
        this._markers.set(marker, markerPromise);
    };
    /**
     * @param {?} marker
     * @return {?}
     */
    MarkerManager.prototype.showBalloon = /**
     * @param {?} marker
     * @return {?}
     */
    function (marker) {
        this.getNativeMarker(marker).then(function (m) {
            m.balloon.open();
        });
    };
    /**
     * @param {?} marker
     * @return {?}
     */
    MarkerManager.prototype.getNativeMarker = /**
     * @param {?} marker
     * @return {?}
     */
    function (marker) {
        return this._markers.get(marker);
    };
    /**
     * @template T
     * @param {?} eventName
     * @param {?} marker
     * @return {?}
     */
    MarkerManager.prototype.createEventObservable = /**
     * @template T
     * @param {?} eventName
     * @param {?} marker
     * @return {?}
     */
    function (eventName, marker) {
        var _this = this;
        return Observable.create(function (observer) {
            _this._markers.get(marker).then(function (m) {
                m.events.add(eventName, function (e) { return _this._zone.run(function () { return observer.next(e); }); });
            });
        });
    };
    MarkerManager.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    MarkerManager.ctorParameters = function () { return [
        { type: YaMapsAPIWrapper },
        { type: NgZone }
    ]; };
    return MarkerManager;
}());
export { MarkerManager };
if (false) {
    /**
     * @type {?}
     * @private
     */
    MarkerManager.prototype._markers;
    /**
     * @type {?}
     * @private
     */
    MarkerManager.prototype._mapsWrapper;
    /**
     * @type {?}
     * @private
     */
    MarkerManager.prototype._zone;
}
