// yamapng/esm5/lib/ya-maps-types.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var yandex;
/**
 * @record
 */
export function YandexMap() { }
if (false) {
    /** @type {?} */
    YandexMap.prototype.events;
    /** @type {?} */
    YandexMap.prototype.geoObjects;
    /**
     * @param {?} param
     * @return {?}
     */
    YandexMap.prototype.setCenter = function (param) { };
    /**
     * @return {?}
     */
    YandexMap.prototype.getCenter = function () { };
    /**
     * @param {?} points
     * @param {?} options
     * @return {?}
     */
    YandexMap.prototype.panTo = function (points, options) { };
    /**
     * @param {?} overlay
     * @return {?}
     */
    YandexMap.prototype.removeOverlay = function (overlay) { };
}
/**
 * @record
 */
export function LatLng() { }
if (false) {
    /**
     * @return {?}
     */
    LatLng.prototype.lat = function () { };
    /**
     * @return {?}
     */
    LatLng.prototype.lng = function () { };
}
/**
 * @record
 */
export function PanToObjects() { }
if (false) {
    /** @type {?} */
    PanToObjects.prototype.points;
    /** @type {?} */
    PanToObjects.prototype.params;
}
/**
 * @record
 */
export function Marker() { }
if (false) {
    /** @type {?} */
    Marker.prototype.events;
    /** @type {?} */
    Marker.prototype.balloon;
    /** @type {?} */
    Marker.prototype.layoutBalloon;
    /** @type {?} */
    Marker.prototype.balloonContentHeader;
    /** @type {?} */
    Marker.prototype.balloonContentBody;
    /** @type {?} */
    Marker.prototype.balloonContentFooter;
    /** @type {?} */
    Marker.prototype.draggable;
    /** @type {?} */
    Marker.prototype.iconLayout;
    /** @type {?} */
    Marker.prototype.iconImageHref;
    /** @type {?} */
    Marker.prototype.iconImageSize;
    /** @type {?} */
    Marker.prototype.iconImageOffset;
    /**
     * @param {?} map
     * @return {?}
     */
    Marker.prototype.setMap = function (map) { };
    /**
     * @param {?} latLng
     * @return {?}
     */
    Marker.prototype.setPosition = function (latLng) { };
}
/**
 * @record
 */
export function Claster() { }
if (false) {
    /**
     * @return {?}
     */
    Claster.prototype.constructor = function () { };
}
/**
 * @record
 */
export function MarkerClaster() { }
if (false) {
    /** @type {?} */
    MarkerClaster.prototype.lat;
    /** @type {?} */
    MarkerClaster.prototype.lng;
    /** @type {?} */
    MarkerClaster.prototype.balloonContentHeader;
    /** @type {?} */
    MarkerClaster.prototype.balloonContentBody;
    /** @type {?} */
    MarkerClaster.prototype.balloonContentFooter;
    /** @type {?} */
    MarkerClaster.prototype.type;
}
/**
 * @record
 */
export function LatLngLiteral() { }
if (false) {
    /** @type {?} */
    LatLngLiteral.prototype.lat;
    /** @type {?} */
    LatLngLiteral.prototype.lng;
}
/**
 * @record
 */
export function MVCObject() { }
if (false) {
    /**
     * @param {?} eventName
     * @param {?} handler
     * @return {?}
     */
    MVCObject.prototype.addListener = function (eventName, handler) { };
}
/**
 * @record
 */
export function MapsEventListener() { }
if (false) {
    /**
     * @return {?}
     */
    MapsEventListener.prototype.remove = function () { };
}
/**
 * @record
 */
export function MouseEvent() { }
if (false) {
    /** @type {?} */
    MouseEvent.prototype.latLng;
    /** @type {?} */
    MouseEvent.prototype.get;
}
/**
 * @record
 */
export function MapClickMouseEvent() { }
if (false) {
    /** @type {?} */
    MapClickMouseEvent.prototype.lat;
    /** @type {?} */
    MapClickMouseEvent.prototype.lng;
}
/**
 * @record
 */
export function MapMouseEvent() { }
if (false) {
    /** @type {?} */
    MapMouseEvent.prototype.lat;
    /** @type {?} */
    MapMouseEvent.prototype.lng;
    /** @type {?} */
    MapMouseEvent.prototype.nativeMarker;
}
/**
 * @record
 */
export function MapOptions() { }
if (false) {
    /** @type {?|undefined} */
    MapOptions.prototype.zoom;
    /** @type {?|undefined} */
    MapOptions.prototype.minZoom;
    /** @type {?|undefined} */
    MapOptions.prototype.maxZoom;
    /** @type {?|undefined} */
    MapOptions.prototype.center;
    /** @type {?} */
    MapOptions.prototype.type;
    /** @type {?|undefined} */
    MapOptions.prototype.controls;
}
/**
 * @record
 */
export function ObjectManager() { }
if (false) {
    /** @type {?} */
    ObjectManager.prototype.clusterize;
    /** @type {?} */
    ObjectManager.prototype.objects;
}
